### PR TITLE
Add reasons for kick/ban + getBanReasons

### PR DESCRIPTION
**Pull request by chrislo27**
### Prerequisites
* [x] I have read and understood the [contribution guidelines](CONTRIBUTING.md)
* [x] This pull request follows the code style of the project
* [x] I have tested this feature thoroughly
  * [Proof of testing](http://i.imgur.com/rl7IQtl.png)

### Changes Proposed in this Pull Request
* Add reason overloads to `IGuild#kickUser` and `IGuild#banUser`
  * I opted to not include them for the deprecated string overloads
* Add `BanReason`
  * Add `IGuild#getBanReasons`

### CODE REVIEW

File: src/main/java/sx/blah/discord/util/BanReason.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package sx.blah.discord.util;
+
+import sx.blah.discord.handle.obj.IGuild;
+import sx.blah.discord.handle.obj.IUser;
+
+/**
+ * Represents a banned user from a guild, with a nullable reason.
+ */
+public class BanReason {
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/util/BanReason.java]: I'm curious, do audit logs use a similar data structure to this?

Comment (chrislo27) [inline src/main/java/sx/blah/discord/util/BanReason.java]: accept my pull request already

Comment (austinv11) [inline src/main/java/sx/blah/discord/util/BanReason.java]: But we gotta make sure we abstract right!

File: src/main/java/sx/blah/discord/handle/impl/obj/Guild.java [hunk 2]
```diff
@@ -395,41 +397,71 @@ public IRole createRole() {
...
 		}
-
-		banUser(userID, 0);
-	}
-
-	@Override
-	public void banUser(long userID, int deleteMessagesForDays) {
-		((DiscordClientImpl) client).REQUESTS.PUT.makeRequest(DiscordEndpoints.GUILDS + id + "/bans/" + Long.toUnsignedString(userID) + "?delete-message-days=" + deleteMessagesForDays);
+		try {
+			((DiscordClientImpl) client).REQUESTS.PUT.makeRequest(DiscordEndpoints.GUILDS + id + "/bans/" + Long.toUnsignedString(userID) + "?delete-message-days=" + deleteMessagesForDays + (reason == null ? "" : ("&reason=" +
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Guild.java]: Unsafe id concat!

File: src/main/java/sx/blah/discord/handle/impl/obj/Guild.java [hunk 3]
```diff
@@ -440,8 +472,17 @@ public void pardonUser(long userID) {
...
 	public void kickUser(IUser user) {
+		kickUser(user, null);
+	}
+
+	@Override
+	public void kickUser(IUser user, String reason) {
 		DiscordUtils.checkPermissions(client, this, user.getRolesForGuild(this), EnumSet.of(Permissions.KICK));
-		((DiscordClientImpl) client).REQUESTS.DELETE.makeRequest(DiscordEndpoints.GUILDS+id+"/members/"+user.getStringID());
+		try {
+			((DiscordClientImpl) client).REQUESTS.DELETE.makeRequest(DiscordEndpoints.GUILDS+id+"/members/"+user.getStringID() + (reason == null ? "" : ("?reason=" + URLEncoder.encode(reason, "UTF-8"))));
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Guild.java]: Unsafe id concat!

File: src/main/java/sx/blah/discord/util/BanReason.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	public BanReason(IGuild guild, IUser user, String reason) {
+		this.guild = guild;
+		this.user = user;
+		this.reason = reason;
+	}
+
+	/**
+	 * Returns the reason for this ban. May be null to signify no reason provided.
+	 * @return The reason, or null if there isn't one
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/util/BanReason.java]: Newline between desc and @return!!!

**Comment by MinnDevelopment** [comment]
Bots will be able to add reasons through the `X-Audit-Log-Reason` header soon, this can be done with any auditable endpoint such as kick/ban/message delete and more.

**Pull request merged**